### PR TITLE
Decouple genericapiserver setDefault from New()

### PR DESCRIPTION
**Pull request by sttts**
---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33575)

### CODE REVIEW

File: pkg/genericapiserver/config.go [hunk 1]
```diff
@@ -255,6 +256,8 @@ func (c *Config) setDefaults() {
 		}
 		c.ExternalHost = hostAndPort
 	}
+
+	return (*defaultedConfig)(&c)
```
Comment (sttts) [inline pkg/genericapiserver/config.go]: A bit ugly

File: pkg/master/master.go [hunk 2]
```diff
@@ -194,17 +194,15 @@ type RESTStorageProvider interface {
...
 
-	gc := *c.Config                                              // copy before mutations
-	gc.EnableSwaggerUI = gc.EnableSwaggerUI && c.EnableUISupport // disable swagger UI if general UI supports it
-	s, err := gc.New()
-	if err != nil {
-		return nil, err
-	}
+	c.EnableSwaggerUI = c.EnableSwaggerUI && c.EnableUISupport // disable swagger UI if general UI supports it
+	withDefaults := c.WithDefaults()
+	c.Config = *(*genericapiserver.Config)(withDefaults) // write back to use defaults below
```
Comment (sttts) [inline pkg/master/master.go]: not the prettiest code either.

**Comment by sttts** [comment]
@liggitt @deads2k wdyt about the defaultedConfig?

File: pkg/genericapiserver/config.go [hunk 3]
```diff
@@ -279,13 +282,11 @@ func (c *Config) setDefaults() {
 //   If the caller wants to add additional endpoints not using the GenericAPIServer's
 //   auth, then the caller should create a handler for those endpoints, which delegates the
 //   any unhandled paths to "Handler".
-func (c Config) New() (*GenericAPIServer, error) {
+func (c *defaultedConfig) New() *GenericAPIServer {
 	if c.Serializer == nil {
-		return nil, fmt.Errorf("Genericapiserver.New() called with config.Serializer == nil")
+		glog.Fatal("Genericapiserver.New() called with config.Serializer == nil")
```
Comment (deads2k) [inline pkg/genericapiserver/config.go]: I don't think I want this fatal.

**Comment by deads2k** [comment]
I don't think this operation is "defaulting".  It's "add missing bits of the config". 

I like the idea of having `New` do nothing but `New` and error if it doesn't have enough information, but I think I'd prefer to have a `func (c config) Complete() (config)`.

File: pkg/genericapiserver/config.go (hunk 3 above)
Comment (sttts) [inline pkg/genericapiserver/config.go]: Isn't that fatal now as well, only with a "failed to launch master: " prefix?

Comment (deads2k) [inline pkg/genericapiserver/config.go]: > Isn't that fatal now as well, only with a "failed to launch master: " prefix?

Callers get a choice.  This removes that choice.

**Comment by sttts** [comment]
@liggitt @deads2k ptal. Renamed `setDefaults` to `Complete` with side-effects, as discussed w/ @deads2k. I left the intermediate type in a third commit. I prefer it that way as it is type-safe. But I don't feel strongly about it. I let you two decide.

File: pkg/master/master.go [hunk 4]
```diff
@@ -194,14 +194,13 @@ type RESTStorageProvider interface {
 // Certain config fields will be set to a default value if unset.
 // Certain config fields must be specified, including:
 //   KubeletClient
-func New(c *Config) (*Master, error) {
+func (c *Config) New() (*Master, error) {
```
Comment (deads2k) [inline pkg/master/master.go]: Add a complete func for this too please.

**Comment by deads2k** [comment]
> @liggitt @deads2k ptal. Renamed setDefaults to Complete with side-effects, as discussed w/ @deads2k. I left the intermediate type in a third commit. I prefer it that way as it is type-safe. But I don't feel strongly about it. I let you two decide.

I prefer having a single type. `Complete` is logically optional if composers may not want to use it.

**Comment by sttts** [comment]
@deads2k if it is only about giving composers the option to ignore it, that's easy:
```go
func (c *Config) SkipCompletion() *completedConfig { return c }
```
That's type-safe and explicit.

File: pkg/master/master.go (hunk 4 above)
Comment (sttts) [inline pkg/master/master.go]: done

**Comment by deads2k** [comment]
> @deads2k if it is only about giving composers the option to ignore it, that's easy:

And I don't feel the need for a second type.

**Comment by sttts** [comment]
@deads2k I have to take that back that I don't care strongly about the intermediate type. I have just spent half an hour to find and fix a missing `Complete` call in some test leading to a panic (b/c nil is not a supported RequestContextMapper). Let's make this type-safe. Go has types which make this easy (compare the last commit). Let's use it.

**Comment by deads2k** [comment]
> @deads2k I have to take that back that I don't care strongly about the intermediate type. I have just spent half an hour to find and fix a missing Complete call in some test leading to a panic (b/c nil is not a supported RequestContextMapper). Let's make this type-safe. Go has types which make this easy (compare the last commit). Let's use it.

I hate anonymous inclusion so much, but I want the change badly enough that I won't force the issue.

File: pkg/master/master.go [hunk 5]
```diff
@@ -116,7 +116,7 @@ const (
 )
 
 type Config struct {
-	*genericapiserver.Config
+	GenericConfig *genericapiserver.Config
```
Comment (deads2k) [inline pkg/master/master.go]: yay.

**Comment by deads2k** [comment]
lgtm.  squash and tag at will.

**Pull request merged**